applications/dashboard/models/class.usermodel.php

```diff
@@ -2563,6 +2563,7 @@ public function search($filter, $orderFields = '', $orderDirection = 'asc', $lim
         // Check for an IP address.
         if (preg_match('`\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}`', $keywords)) {
             $ipAddress = $keywords;
+            $optimize = false;
```

Comment:
Unless this is part of the fix, I recommend not including it. There's the possibility future updates to this method that rely on the value `$optimize` could unintentionally be dodged when searching for an IP. If "optimizing" the query isn't causing or contributing to the issue, it can probably stay as is.

Reply:
This is the fix.  If optimize is true then we append ** and u.Name like :uName
 and u.Deleted = :uDeleted** to the query in and in this case the keyword would be the ip.

I was thinking doing something like this 

 if ($optimize && !isset($ipAddress)) {
  //do optimize
}

thoughts?

Reply:
> This is the fix. If optimize is true then we append ** and u.Name like :uName
and u.Deleted = :uDeleted** to the query in and in this case the keyword would be the ip.

Isn't the un-optimized query also using the IP address for Name and Email matches?

https://github.com/vanilla/vanilla/blob/017304c3f1a07a8e7bdd291dd3bfd08cec49951f/applications/dashboard/models/class.usermodel.php#L2608-L2615

Fixing the multi-dimensional array issue appears to be what actually fixed this. However, there do appear to be deeper issues with the method. I believe the reason fixing the multi-dimensional array issue works is due to MySQL's [order of operations](https://dev.mysql.com/doc/refman/5.7/en/operator-precedence.html). An "optimized" user search IP query may end up generating the following:

```mysql
select u.*
from GDN_User u
where u.UserID in (1, 2, 3, 4, 5)
 or u.LastIPAddress in ('192.168.1.1', [packed version of the IP])
 and u.Name like '192.168.1.1%';
```

1. The AND is executed, first: is the user's IP one we're looking for AND is their name the IP address? No.
1. The OR is executed, next: is the user's ID one we're looking for OR (result from the AND condition)? Yes.

Prior to this fix, all conditions were failing, because the user's ID wasn't in the list. The list was nulled out, so the condition was false.

You can see this in action. Try running:

```mysql
select *
from GDN_User
where true or true and false
```

versus

```mysql
select *
from GDN_User
where false or true and false
```

The un-optimized query would generated something like:

```mysql
select u.*
from GDN_User u
where u.UserID in (1, 2, 3, 4, 5)
 or u.LastIPAddress in ('192.168.1.1', [packed version of the IP])
 or (u.Name like '192.168.1.1%' or u.Email like '192.168.1.1%');
```

There is no AND. If any of the above are true, the user makes the cut. This can be seen as:

```mysql
select *
from GDN_User
where true or true or false
```